xls/examples/lfsr.x

```diff
@@ -0,0 +1,518 @@
+////////////////////////////////////////////////////////////////////////////////
+// Linear-feedback shift register (LFSR)
+// A template function to compute the next value of an LFSR counter.
+// Taps can be specified to have varying periods for a given number of bits.
+//
+// For example, lfsr(u5:4, u5:0b10100) computes the value that comes after 4
+// in a 5-bit LFSR with taps on bits 2 and 4.
+////////////////////////////////////////////////////////////////////////////////
+
+fn lfsr<BIT_WIDTH: u32>(current_value: uN[BIT_WIDTH], tap_mask: uN[BIT_WIDTH]) -> uN[BIT_WIDTH] {
+	// Compute the new bit from the taps
+	let new_bit = for (index, xor_bit): (u32, u1) in range(u32:0, BIT_WIDTH) {
+		if tap_mask[index+:u1] == u1:0 {
+			xor_bit
+		} else {
+			xor_bit ^ current_value[index+:u1]
+		}
+	} (u1:0);
+
+	// Kick the high bit and insert the new bit
+	current_value[0:{(BIT_WIDTH - u32:1) as s32}] ++ new_bit
+}
+
+////////////////////////////////////////////////////////////////////////////////
+// A series of maximal LFSRs for different bit widths.
+// Source: https://en.wikipedia.org/wiki/Linear-feedback_shift_register
+////////////////////////////////////////////////////////////////////////////////
+
+fn lfsr2(n: u2) -> u2 {
```

Comment:
not sure if you need those wrapper, the interpreter should be able to infer the function type parameters.

Reply:
ah I see, so maybe we can only keep a few ones that are actually tested and and a descriptive comment (X-bit lfsr with a period of XXX)?

also curious if it's possible to predict the period from the tap masks?

Reply:
No, you cannot predict the period from the tap masks. All you know is that the maximal theoretical period is 2^n - 1.
I think all you can do is try all kind of masks until you find a maximal one.
For a low number of bits, it is fast enough to try all combination and find a sequence that has the exact period you want. Assuming such a sequence exists of course.
I think Xilinx has discovered a maximal LFSR on 168 bits. It is the longest I have heard of, and it uses only 4 tap bits.

Reply:
Thanks for the detailed explanation!

Would it make sense to capture some of those "famous" LFSR in the examples?

Reply:
I removed  the superfluous examples, keeping only the tested ones.
I also added succint comments about the number of bits and period.

Reply:
So, I actually tried to test the period for a 24-bit LFSR, but the execution time was horrenduous.
A cc_test would probably fair better, but in the interpreter, it is just not pratical.
To assert the period, one can start from a given value (let's say 1), make a full cycle (by calling the function "period" times) and finally check that the last yielded value is the starting one (i.e. 1). Additionally, one would probably also test that the LFSR does not only return the same value all the time.

Reply:
I'm not sure if the `cc_test` would be faster: it would just help to generate the value from a reference implementation rather than hardcoding them; but it would still use the same interpreted code under the hood (unless you use the JIT!)

Reply:
we could leverage the JIT for this to be faster, but this could be done in a separate PR.